systems/controllers/osc/operational_space_control.cc

```diff
@@ -0,0 +1,619 @@
+  int n_u_w_spr = tree_w_spr.get_num_actuators();
+  cout << "n_q_w_spr = " << n_q_w_spr << endl;
+  cout << "n_v_w_spr = " << n_v_w_spr << endl;
+  cout << "n_u_w_spr = " << n_u_w_spr << endl;
+
+  // Input/Output Setup
+  state_port_ = this->DeclareVectorInputPort(
+      OutputVector<double>(n_q_w_spr, n_v_w_spr, n_u_w_spr)).get_index();
+  this->DeclareVectorOutputPort(TimestampedVector<double>(n_u_w_spr),
+                                &OperationalSpaceControl::CalcOptimalInput);
+  if (used_with_finite_state_machine) {
+    fsm_port_ = this->DeclareVectorInputPort(
+                  BasicVector<double>(1)).get_index();
+
+    // Discrete update to record the last state event time
+    DeclarePerStepDiscreteUpdateEvent(
+        &OperationalSpaceControl::DiscreteVariableUpdate);
+    prev_fsm_state_idx_ = this->DeclareDiscreteState(-0.1 * VectorXd::Ones(1));
+    prev_event_time_idx_ = this->DeclareDiscreteState(VectorXd::Zero(1));
+  }
+
+  // Initialize the mapping from spring to no spring
+  map_position_from_spring_to_no_spring_ = MatrixXd::Zero(n_q_, n_q_w_spr);
+  for (int i = 0; i < n_q_; i++) {
+    bool successfully_added = false;
+    for (int j = 0; j < n_q_w_spr; j++) {
+      std::string name_wo_spr = tree_wo_spr_.get_position_name(i);
+      std::string name_w_spr = tree_w_spr_.get_position_name(j);
+      if (name_wo_spr.compare(0, name_wo_spr.size(), name_w_spr) == 0) {
+        map_position_from_spring_to_no_spring_(i, j) = 1;
+        successfully_added = true;
+      }
+    }
+    DRAKE_DEMAND(successfully_added);
+  }
+  map_velocity_from_spring_to_no_spring_ = MatrixXd::Zero(n_v_, n_v_w_spr);
+  for (int i = 0; i < n_v_; i++) {
+    bool successfully_added = false;
+    for (int j = 0; j < n_v_w_spr; j++) {
+      std::string name_wo_spr = tree_wo_spr_.get_velocity_name(i);
+      std::string name_w_spr = tree_w_spr_.get_velocity_name(j);
+      if (name_wo_spr.compare(0, name_wo_spr.size(), name_w_spr) == 0) {
+        map_velocity_from_spring_to_no_spring_(i, j) = 1;
+        successfully_added = true;
+      }
+    }
+    DRAKE_DEMAND(successfully_added);
+  }
+
+  // Get input limits
+  VectorXd u_min(n_u_);
+  VectorXd u_max(n_u_);
+  for (int i = 0; i < n_u_; i++) {
+    u_min(i) = tree_wo_spr_.actuators[i].effort_limit_min_;
+    u_max(i) = tree_wo_spr_.actuators[i].effort_limit_max_;
+  }
+  u_min_ = u_min;
+  u_max_ = u_max;
+
+  // Check if the model is floating based
+  is_quaternion_ = multibody::IsFloatingBase(tree_w_spr);
+}
+
+// Cost methods
+void OperationalSpaceControl::AddAccelerationCost(std::string joint_vel_name,
+                                                  double w) {
+  if (W_joint_accel_.size() == 0) {
+    W_joint_accel_ = Eigen::MatrixXd::Zero(n_v_, n_v_);
+  }
+  int idx = makeNameToVelocitiesMap(tree_wo_spr_).at(joint_vel_name);
+  W_joint_accel_(idx, idx) += w;
+}
+
+// Constraint methods
+void OperationalSpaceControl::AddContactPoint(std::string body_name,
```

Comment:
Same comment here as for the tracking data. Won't this all get out of sync if both `AddContactPoint` and `AddStateAndContactPoint` are called? A better way of tracking "all states" is needed than emptiness of the state vector.

Reply:
There is a private method `CheckConstraintSettings` that checks if the state length is correct.